framework/src/test/java/io/github/kgress/scaffold/page/BasePageTests.java

```diff
@@ -88,6 +90,28 @@ public void verifyIsOnPage_pageNotLoaded() {
         assertTrue(exception.getMessage().contains("The intended page failed to load"));
     }
 
+    @Test
+    public void waitUntilPageIsLoaded_targetFrameDetachedThrowsAfterFiveTries() {
+        var spyAutomationWait = spy(new AutomationWait(mockWebDriverWrapper, 1L));
+        doThrow(new WebDriverException("target frame detached"))
+                .when(spyAutomationWait)
+                .waitForCustomCondition(any(), any());
+        assertThrows(WebDriverException.class, spyAutomationWait::waitUntilPageIsLoaded);
+    }
+
+    @Test
+    public void waitUntilPageIsLoaded_targetFrameDetachedDoesNotThrowUnderFiveTries() {
+        var spyAutomationWait = spy(new AutomationWait(mockWebDriverWrapper, 1L));
+        doThrow(new WebDriverException("target frame detached"))
+                .doThrow(new WebDriverException("target frame detached"))
+                .doThrow(new WebDriverException("target frame detached"))
+                .doThrow(new WebDriverException("target frame detached"))
+                .doReturn(true)
+                .when(spyAutomationWait)
+                .waitForCustomCondition(any(), any());
+        assertTrue(spyAutomationWait.waitUntilPageIsLoaded());
+    }
+
```

Comment:
Could you change `waitUntilPageIsLoaded()` to be package private, instead of private? Then you could test that method by invoking it directly and setting it to a variable. I don't think the spy is actually allowing you to execute real code, instead of just interacting with a mock. 

Generally speaking, the flow would be to mock your dependencies, and then interact with the real code. The correct thing is being mocked (in this case, the automation wait), but we're just missing the true interaction.

Unless I'm not understanding the spy in the correct way.

Reply:
Spy does the default action unless you specify mocked behavior, in this case I wanted the waitForCustomCondition() to throw the exception and then run the real code via `spyAutomationWait.waitUntilPageIsLoaded()`. Using the mockAutomationWait causes all of the methods to in AutomationWait to be mocked instead of just the `waitForCustomCondition` I needed to test this.

You can verify the behavior due to the amount of time the test takes, since `waitUntilPageIsLoaded_targetFrameDetachedDoesNotThrowUnderFiveTries` runs the `AutomationUtils.sleep(200);` code 4 times it takes roughly 900ms while
 `waitUntilPageIsLoaded_targetFrameDetachedThrowsAfterFiveTries` runs five times, taking 1100ms:

<img width="632" alt="Screen Shot 2022-05-04 at 4 37 33 PM" src="https://user-images.githubusercontent.com/95888307/166829938-c6c51ba4-7a98-4ba1-85f2-ed3c394c4b10.png">

Reply:
Ah, I see. `Spy` doesn't mock the entire object, but granularly mocks methods.